scripts/poe/lib/validate.js

```diff
@@ -34,16 +34,16 @@ const validDirectory = (fpath) => {
 const validatePlaceHolders = (langs, dir) => {
   let valid = true;
   const templateFile = `${dir}/messages-en.properties`;
-  const templatePlaceholders = extractPlaceholders(templateFile);
+  const templatePlaceholders = extractPlaceholders(templateFile, true);
   langs.filter(lang => lang !== 'en').forEach(lang => {
     const file = `${dir}/messages-${lang}.properties`;
-    const placeholders = extractPlaceholders(file);
-    Object.keys(placeholders).forEach(k => {
-      const placeholder = placeholders[k];
-      const templatePlaceholder = templatePlaceholders[k];
-      if (placeholder.placeholders.toString() !== templatePlaceholder.placeholders.toString()) {
+    const placeholdersWithIndex = extractPlaceholders(file);
+    Object.keys(placeholdersWithIndex).forEach(k => {
+      const placeholderWithIndex = placeholdersWithIndex[k];
+      const foundAllPlaceholders = placeholderWithIndex.placeholders.every(el => templatePlaceholders.includes(el));
```

Comment:
If I'm reading this correctly then if one translation in English has a placeholder `{{test}}` then we will allow all terms to have the placeholder `{{test}}`. I think this is too permissive.

When we translate something in the webapp we pass a different context depending on what we're translating so just because one `context` has `test` doesn't mean the next one will.

This makes it really complicated to work out what are valid placeholders and what aren't, but this only caused one or two failures when I ran it. My idea was to have another file which lists additional valid placeholders for each key. This would be updated manually if the validation starts failing after the developer has checked that it is in fact valid. This is a pain but as it's not common I think it'll be enough for now.

What do you think?

Reply:
Sounds good. Less automated but more effective. Hopefully, there are not many manual updates during releases.